elide-core/src/main/java/com/yahoo/elide/core/EntityDictionary.java

```diff
@@ -603,16 +604,94 @@ public RelationshipType getRelationshipType(Object entity, String relation) {
 
     /**
      * Get a type for a field on an entity.
+     * <p>
+     * If this method is called on a bean such as the following
+     * <pre>
+     * {@code
+     * public class Address {
+     *
+     *     {@literal @}Id
+     *     private Long id
+     *
+     *     private String street1;
+     *
+     *     private String street2;
+     * }
+     * }
+     * </pre>
+     * then
+     * <pre>
+     * {@code
+     * getType(Address.class, "id") = Long.class
+     * getType(Address.class, "street1") = String.class
+     * getType(Address.class, "street2") = String.class
+     * }
+     * </pre>
+     * But if the ID field is not "id" and there is no such non-ID field called "id", i.e.
+     * <pre>
+     * {@code
+     * public class Address {
+     *
+     *     {@literal @}Id
+     *     private Long surrogateKey
+     *
+     *     private String street1;
+     *
+     *     private String street2;
+     * }
+     * }
+     * </pre>
+     * then
+     * <pre>
+     * {@code
+     * getType(Address.class, "id") = Long.class
+     * getType(Address.class, "surrogateKey") = Long.class
+     * getType(Address.class, "street1") = String.class
+     * getType(Address.class, "street2") = String.class
+     * }
+     * </pre>
+     * JSON-API spec does not allow "id" as non-ID field name. If, therefore, there is a non-ID field called "id",
+     * callling this method has undefined behavior
      *
      * @param entityClass Entity class
      * @param identifier  Field to lookup type
      * @return Type of entity
      */
     public Class<?> getType(Class<?> entityClass, String identifier) {
+        if (isIdType(entityClass, identifier)) {
```

Comment:
In your first commit, you had code like:
```java
        if (!ID_IDENTIFIER.equals(getEntityBinding(entityClass).getIdFieldName())) {
            return getIdType(entityClass);
        }```

Is there a reason we have to define a new method `isIdType` or could we use the code from the first commit?

Reply:
yes, with first commit, if `identifier` is a non-ID field and the bean ID is not "id", the first commit returns the type of ID while we are asking for the type of a non-ID field instead

The first commit also has undefined behavior in tests on Beans which do not have ID defined such as `Address.java`. The new method handles that case gracefully.

I can figure out a way to remove the new method if that's needed

Reply:
How about:

```java
if (identifier.equals(ID_IDENTIFIER) {
  return getEntityBinding(entityClass).getIdFieldType();
}
```

Reply:
Yeah looks simpler.